File: ros_tank_py/path_follow_node.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Pose, Point, Twist
from std_msgs.msg import Float32MultiArray, Float32, Bool
import math
import numpy as np
from tf_transformations import euler_from_quaternion


#from ros_tank_py.srv import Waypoint
import numpy as np

class PathFollowNode(Node):

	def __init__(self):
		super().__init__('path_follow_node')
		self.path_sub = self.create_subscription(Float32MultiArray, '/path', self.path_callback, 10)
		self.pose_sub = self.create_subscription(Pose, '/estimated_pose', self.pose_callback, 10)
		self.yaw_tgt_pub = self.create_publisher(Float32, '/yaw_tgt', 10)
		self.path_done_pub = self.create_publisher(Bool, '/path_done', 10)
		self.position = []
		self.orientation = []
		self.pose_rxd = False
		self.path = Float32MultiArray()
		self.lookahead_dist = 0.15
		self.last_point = False
		self.path_rxd = False
		self.path_done = Bool()
		self.target_yaw = Float32()
		self.timer = self.create_timer(0.1, self.timer_callback)

	def pose_callback(self, msg):
		self.position = [msg.position.x,
				msg.position.y,
				msg.position.z]

		orient = msg.orientation
		_, _, self.yaw = euler_from_quaternion([msg.orientation.x,
				msg.orientation.y,
				msg.orientation.z,
				msg.orientation.w])
		self.pose_rxd = True


	def find_lookahead_point(self):
		# Find the closest point on the path
		#distances = np.linalg.norm(self.path - self.position, axis=1)
		#closest_idx = np.argmin(distances)
		last_distance = 10000
		for i in range(len(self.path)):
			distance = np.linalg.norm(self.path[i] - self.position)
			if distance > last_distance:
				break
			closest_idx = i
			#self.get_logger().info(f"{closest_idx}")
			last_distance = distance
		if closest_idx > 0:
			self.path = self.path[closest_idx-1:]
		#self.get_logger().info(f"Path Length: {len(self.path)}")

		# Start searching from the closest point onward
		for i in range(len(self.path)):
			dist = np.linalg.norm(self.path[i] - self.position)
			if dist >= self.lookahead_dist:
				# Found the lookahead point
				return self.path[i], False

		# If no point is found beyond lookahead distance, return the last point
		return self.path[-1], True

	def path_callback(self, msg):
		"""Callback for receiving the path as Float32MultiArray"""
		# The path is expected to be a sequence of [x1,y1,z1, x2,y2,z2, ...]
		if not self.path_rxd:
			data = np.array(msg.data)
			self.path = data.reshape(-1, 3)  # Reshape into Nx3 array
			self.path_rxd = True
			#self.get_logger().info(f"Received new path with {len(self.path)} points")

	def yaw_limit(self):
		if self.yaw_tgt.data  < -math.pi:
			self.yaw_tgt.data += 2 * math.pi
		elif self.yaw_tgt.data > math.pi:
			self.yaw_tgt.data -= 2 * math.pi

	def calculate_yaw_tgt(self, lookahead_point):
		if lookahead_point is None:
			return None

		dx = lookahead_point[0] - self.position[0]
		dy = lookahead_point[1] - self.position[1]

		target_yaw = math.atan2(dy, dx)

		return target_yaw

	def path_follow(self):
		if not self.path_rxd:
			self.get_logger().info(f"No Path!")
			return

		if not self.pose_rxd:
			self.get_logger().info(f"No Pose!")
			return

		if self.path_done.data:
			self.path_done_pub.publish(self.path_done)
			return

		lookahead_point, last_point = self.find_lookahead_point()
		self.get_logger().info(f"Path done: {last_point}")
		self.target_yaw.data = self.calculate_yaw_tgt(lookahead_point)

		if lookahead_point is None:
			return
		self.get_logger().info(f"Lookahead Point: {lookahead_point[0]}, {lookahead_point[1]}")
		#self.get_logger().info(f"Yaw Target: {self.target_yaw}")

		self.path_done.data = last_point
		self.yaw_tgt_pub.publish(self.target_yaw)
		self.path_done_pub.publish(self.path_done)

	def timer_callback(self):
		#self.get_logger().info('Following Path')
		self.path_follow()

def main(args=None):
	rclpy.init(args=args)
	path_follow_node = PathFollowNode()
	rclpy.spin(path_follow_node)
	path_follow_node.destroy_node()
	rclpy.shutdown()

if __name__ == '__main__':
	main()

File: ros_tank_cpp/path_follow_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <std_msgs/msg/float32.hpp>
#include <std_msgs/msg/bool.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <cmath>
#include <vector>
#include <limits>

class PathFollowNode : public rclcpp::Node
{
public:
    PathFollowNode() : Node("path_follow_node")
    {
        // Create subscriptions
        path_sub_ = this->create_subscription<std_msgs::msg::Float32MultiArray>(
            "/path", 10, std::bind(&PathFollowNode::path_callback, this, std::placeholders::_1));
        
        pose_sub_ = this->create_subscription<geometry_msgs::msg::Pose>(
            "/estimated_pose", 10, std::bind(&PathFollowNode::pose_callback, this, std::placeholders::_1));

        // Create publishers
        yaw_tgt_pub_ = this->create_publisher<std_msgs::msg::Float32>("/yaw_tgt", 10);
        path_done_pub_ = this->create_publisher<std_msgs::msg::Bool>("/path_done", 10);

        // Initialize variables
        position_.resize(3, 0.0);
        orientation_.resize(4, 0.0);
        pose_rxd_ = false;
        lookahead_dist_ = 0.15;
        last_point_ = false;
        path_rxd_ = false;
        yaw_ = 0.0;

        // Create timer (10 Hz)
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100), std::bind(&PathFollowNode::timer_callback, this));
    }

private:
    void pose_callback(const geometry_msgs::msg::Pose::SharedPtr msg)
    {
        position_[0] = msg->position.x;
        position_[1] = msg->position.y;
        position_[2] = msg->position.z;

        // Convert quaternion to euler angles
        tf2::Quaternion quat(
            msg->orientation.x,
            msg->orientation.y,
            msg->orientation.z,
            msg->orientation.w
        );
        
        tf2::Matrix3x3 mat(quat);
        double roll, pitch;
        mat.getRPY(roll, pitch, yaw_);
        
        pose_rxd_ = true;
    }

    std::pair<std::vector<double>, bool> find_lookahead_point()
    {
        // Find the closest point on the path
        double last_distance = 10000.0;
        int closest_idx = 0;
        
        for (int i = 0; i < static_cast<int>(path_.size()); i++)
        {
            double dx = path_[i][0] - position_[0];
            double dy = path_[i][1] - position_[1];
            double dz = path_[i][2] - position_[2];
            double distance = std::sqrt(dx*dx + dy*dy + dz*dz);
            
            if (distance > last_distance)
            {
                break;
            }
            closest_idx = i;
            last_distance = distance;
        }
        
        if (closest_idx > 0)
        {
            // Remove points before closest_idx-1
            path_.erase(path_.begin(), path_.begin() + closest_idx - 1);
        }

        // Start searching from the closest point onward
        for (int i = 0; i < static_cast<int>(path_.size()); i++)
        {
            double dx = path_[i][0] - position_[0];
            double dy = path_[i][1] - position_[1];
            double dz = path_[i][2] - position_[2];
            double dist = std::sqrt(dx*dx + dy*dy + dz*dz);
            
            if (dist >= lookahead_dist_)
            {
                // Found the lookahead point
                return std::make_pair(path_[i], false);
            }
        }

        // If no point is found beyond lookahead distance, return the last point
        return std::make_pair(path_.back(), true);
    }

    void path_callback(const std_msgs::msg::Float32MultiArray::SharedPtr msg)
    {
        // The path is expected to be a sequence of [x1,y1,z1, x2,y2,z2, ...]
        if (!path_rxd_)
        {
            path_.clear();
            
            // Reshape data into Nx3 array
            for (size_t i = 0; i < msg->data.size(); i += 3)
            {
                if (i + 2 < msg->data.size())
                {
                    std::vector<double> point = {
                        msg->data[i],
                        msg->data[i + 1],
                        msg->data[i + 2]
                    };
                    path_.push_back(point);
                }
            }
            
            path_rxd_ = true;
            // RCLCPP_INFO(this->get_logger(), "Received new path with %lu points", path_.size());
        }
    }

    double normalize_angle(double angle)
    {
        while (angle < -M_PI)
        {
            angle += 2 * M_PI;
        }
        while (angle > M_PI)
        {
            angle -= 2 * M_PI;
        }
        return angle;
    }

    double calculate_yaw_tgt(const std::vector<double>& lookahead_point)
    {
        if (lookahead_point.empty())
        {
            return 0.0;
        }

        double dx = lookahead_point[0] - position_[0];
        double dy = lookahead_point[1] - position_[1];

        double target_yaw = std::atan2(dy, dx);

        return target_yaw;
    }

    void path_follow()
    {
        if (!path_rxd_)
        {
            RCLCPP_INFO(this->get_logger(), "No Path!");
            return;
        }

        if (!pose_rxd_)
        {
            RCLCPP_INFO(this->get_logger(), "No Pose!");
            return;
        }

        if (path_done_.data)
        {
            path_done_pub_->publish(path_done_);
            return;
        }

        if (path_.empty())
        {
            return;
        }

        auto [lookahead_point, last_point] = find_lookahead_point();
        //RCLCPP_INFO(this->get_logger(), "Path done: %s", last_point ? "true" : "false");
        
        double target_yaw = calculate_yaw_tgt(lookahead_point);

        if (lookahead_point.empty())
        {
            return;
        }
        
        //RCLCPP_INFO(this->get_logger(), "Lookahead Point: %.3f, %.3f", 
        //            lookahead_point[0], lookahead_point[1]);

        // Normalize target yaw
        target_yaw = normalize_angle(target_yaw);

        path_done_.data = last_point;
        
        // Publish messages
        auto target_yaw_msg = std_msgs::msg::Float32();
        target_yaw_msg.data = target_yaw;
        yaw_tgt_pub_->publish(target_yaw_msg);
        path_done_pub_->publish(path_done_);
    }

    void timer_callback()
    {
        // RCLCPP_INFO(this->get_logger(), "Following Path");
        path_follow();
    }

    // Subscriptions
    rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr path_sub_;
    rclcpp::Subscription<geometry_msgs::msg::Pose>::SharedPtr pose_sub_;
    
    // Publishers
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr yaw_tgt_pub_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr path_done_pub_;
    
    // Timer
    rclcpp::TimerBase::SharedPtr timer_;

    // State variables
    std::vector<double> position_;
    std::vector<double> orientation_;
    bool pose_rxd_;
    std::vector<std::vector<double>> path_;
    double lookahead_dist_;
    bool last_point_;
    bool path_rxd_;
    std_msgs::msg::Bool path_done_;
    double yaw_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PathFollowNode>());
    rclcpp::shutdown();
    return 0;
}
